segmentstore/server/host/src/main/java/io/pravega/segmentstore/server/host/handler/PravegaRequestProcessor.java

```diff
@@ -377,101 +376,89 @@ private void handleException(long requestId, String segment, String operation, T
+                .thenAccept(txName -> connection.send(new TransactionCreated(requestId, createTransaction.getSegment(), createTransaction.getTxid())))
+                .exceptionally(e -> handleException(requestId, createTransaction.getSegment(), "Create transaction", e));
     }
 
     @Override
     public void commitTransaction(CommitTransaction commitTx) {
         String transactionName = StreamSegmentNameUtils.getTransactionNameFromId(commitTx.getSegment(), commitTx.getTxid());
         long requestId = commitTx.getRequestId();
         log.debug("Commiting transaction {} ", commitTx);
-        segmentStore.sealStreamSegment(transactionName, TIMEOUT)
-                    .exceptionally(this::ignoreSegmentSealed)
-                    .thenCompose(v -> recordStatForTransaction(transactionName, commitTx.getSegment())
-                                 .exceptionally((Throwable e) -> {
-                                     // gobble up any errors from state recording so we do not affect rest of the flow.
-                                     log.error("exception while computing stats while merging txn {}", e);
-                                     return null;
-                                 }))
-                    .thenApply(result -> {
-                        segmentStore.mergeTransaction(transactionName, TIMEOUT).thenAccept(v -> {
-                            connection.send(new TransactionCommitted(requestId, commitTx.getSegment(),
-                                    commitTx.getTxid()));
-                        }).exceptionally((Throwable e) -> {
-                            handleException(requestId, transactionName, "Commit transaction", e);
+
+        // Seal and Merge can execute concurrently, as long as they are invoked in the correct order (first Seal, then Merge).
+        // If Seal fails for whatever reason (except already sealed), then Merge will also fail because the txn is not sealed,
+        // but invoking them in parallel does provide benefits in terms of reduced latency.
+        val seal = segmentStore
+                .sealStreamSegment(transactionName, TIMEOUT)
+                .exceptionally(this::ignoreSegmentSealed)
+                .thenCompose(v -> recordStatForTransaction(transactionName, commitTx.getSegment())
+                        .exceptionally(e -> {
+                            // gobble up any errors from stat recording so we do not affect rest of the flow.
+                            log.error("exception while computing stats while merging txn {}", e);
                             return null;
-                        });
-                        return null;
-                    })
-                    .exceptionally((Throwable e) -> {
-                        handleException(requestId, transactionName, "Commit transaction", e);
-                        return null;
-                    });
+                        }));
+        val merge = segmentStore
```

Comment:
what could be the consequence of doing thesmerge and seal in parallel?

for example, a bad application code could be writing to a txn while calling commit on it concurrently.. 

if we are merging an unsealed segment, do we run the risk of some unwanted races?

Reply:
You cannot merge an unsealed segment. The OperationProcessor will reject a merge if the segment isn't sealed. If you look, this is all server-side code and is invoked every time someone wants to merge a txn. On the SSS side, we could choose to automatically seal the segment upon a merge, but we left it open, which gives more flexibility. This change has no effect from the outside, except that it will execute faster.